src/psnr.jl

```diff
@@ -0,0 +1,67 @@
+"""
+    PeakSignalNoiseRatio <: FullReferenceIQI
+    psnr(x, ref [, peakval])
+    assess(PSNR(), x, ref, [, peakval])
+
+Peak signal-to-noise ratio (PSNR) is used to measure the quality of image.
+
+PSNR (in dB) is calculated by `10log10(PeakVal^2/mse(x, ref))`, where `PeakVal`
+is the maximum possible pixel value of image `ref`, and
+[`mse`](@ref ImageDistances.mse) is the mean squared error between images `x`
+and `ref`.
+
+!!! info
+
+    For general `Color3` images, PSNR is reported against each channel of that
+    color space as a `Vector`. One exception is `AbstractRGB` images, where
+    `mse` is calculated over three channels by convention.
```

Comment:
A dilemma here is: If we stick to "array elements are pixels", then we should not make `AbstractRGB` as an exception. But most people would be surprised to get a vector result for RGB images.

Concerning to N-d arrays, making such exception would make it hard/impossible to define PSNR for 3d RGB images.

Reply:
For simplicity consideration, I'm planning to make `psnr` outputs a Vector for whatever inputs.

For gray image inputs, it is length-1 vector, for any Color3 image inputs, it is length-3 vector. In the beginning, it would surprise users for a while to get used to `mean(psnr(gray_x, gray_ref))` and `mean(psnr(channelview(rgb_x, rgb_ref)))`, since it's a little lengthy.

Another strategy is to make `psnr` output results that stick to the convention (i.e., output scalar for Gray and RGB images, throw error for other images), while keeping `PSNR()(x, ref)` and `assess(PSNR(), x, ref)` output a vector for each channel of color space. --> I prefer this idea, will change it later.

Reply:
That's a good question

I don't have a better solution here, but `psnr(mappedarray(c->c.l, x, ref))` will use `peakval=1` as default here. That's probably not what we expected. I prefer to throw an error instead of returning incorrect result.

Once we split images into channels, we lost the information of the colorspace, and we can't infer the peakval anymore in a safe way. Manually splitting images into channels will eventually meet this problem, it's just a matter of here or there.

Reply:
In the new commit, I take the following decision:

* If we pass scalar `peakval` (by default it is scalar for Gray and RGB images), then the PSNR output is scalar as well

```julia
x = rand(Gray, 3, 3)
y = rand(Gray, 3, 3)
psnr(x, y) # scalar
psnr(x, y, 1.0) # scalar
```

```julia
x = rand(RGB, 3, 3)
y = rand(RGB, 3, 3)
psnr(x, y) # scalar
psnr(x, y, 1.0) # scalar
```

* If we pass vector `peakval`, then the PSNR is calculated channelwise, the output is vector as well:

```julia
x = rand(Gray, 3, 3)
y = rand(Gray, 3, 3)
psnr(x, y, [1.0, ] ) # vector
```

```julia
x = rand(RGB, 3, 3)
y = rand(RGB, 3, 3)
psnr(x, y, [1.0, 1.0, 1.0]) # vector
```